### PR TITLE
Azure Deep Storage extension

### PR DESCRIPTION
**Pull request by davrodpin**
Adding a new Deep Storage implementation using [Microsoft Azure Storage](http://azure.microsoft.com/en-us/services/storage/).

Common configuration options:

```
druid.storage.type=azure
druid.azure.account=<azure storage account>
druid.azure.key=<azure storage account key>
druid.azure.container=<azure storage container>
druid.azure.protocol=<optional; valid options: https or http; default: https>
druid.azure.maxTries=<optional; number of tries before give up an Azure operation; default: 3; min: 1>
```

### CODE REVIEW

**Comment by drcrallen** [comment]
If my research is correct, Azure does not have a unique URI scheme. This would make this PR the first deep storage impl that is not identifiable by URI alone, but *requires* either non-standard identification wrapping (segment spec) or host capability negotiations. Is there ANYONE out there who uses some sort of URI prefix (as per cassandra using `cassandra://`) to indicate the special handlers needed for Azure connections?

**Comment by davrodpin** [comment]
Hi @drcrallen,

I don't think Azure provides specific URI prefix on its API.

There is a section on this [link](https://msdn.microsoft.com/en-us/library/azure/dd135715.aspx), "Resource URI syntax", that describes the URI syntax for storage resources.

**Comment by drcrallen** [comment]
Note that the lack of support for Azure-specific URI is not a blocker per-se, but it does limit the capacity to implement `io.druid.segment.loading.URIDataPuller` and any future capacities that handle URIs in a smarter manner.

If a `io.druid.segment.loading.URIDataPuller` (and potentially pusher/archiver/whatever in the future) which handles general `https` targets would handle Azure correctly then it should not be an issue. Even if `https` would work but `azure` or some other scheme is more "optimized" that should be ok. But if functionality requires `com.microsoft.azure` packages even over `https`, then this could cause conflicts down the road.

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/CompressionUtilsWrapper.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import com.metamx.common.CompressionUtils;
+import com.metamx.common.FileUtils;
+
+import java.io.File;
+import java.io.IOException;
+
+public class CompressionUtilsWrapper
+{
+
+  public FileUtils.FileCopyResult unzip(
```
Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/CompressionUtilsWrapper.java]: Why is this needed?

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+  @Override
+  public String getPathForHadoop(String dataSource)
+  {
+    return null;
+  }
+
+  public File createCompressedSegmentDataFile(final File indexFilesDir) throws IOException
+  {
+    final File zipOutFile = File.createTempFile(TEMP_DIR, AzureStorageDruidModule.INDEX_ZIP_FILE_NAME);
```
Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: `File. createTempFile` takes prefix / suffix, not dir / Filename. That makes this confusing at first glance.

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/CompressionUtilsWrapper.java (hunk 1 above)
Comment (davrodpin) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/CompressionUtilsWrapper.java]: Only way I could find to add unit tests for AzureDataSegmentPuller once all CompressionUtils methods are static and it make life harder on unit tests.

The only other option that I saw was adding [PowerMock](https://code.google.com/p/powermock/) that allows me to mock static and final classes/methods, but to be frank, I don't like much this idea.

Anyways, I can remove it and also remove the Puller unit test.

Any thoughts?

Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/CompressionUtilsWrapper.java]: My preferred method is to mock the connector as per https://github.com/metamx/druid/blob/queryTimeLookup/extensions/s3-extensions/src/test/java/io/druid/storage/s3/S3DataSegmentPullerTest.java 

If possible, to simply mocking the calls to AzureStorageContainer should yield a more "pure" unit test.

**Comment by drcrallen** [comment]
Overall I'm feeling good about this PR, but would like to see the classes changed to not have an instanced Compression Utils if possible. I would still like to have a second look at the PR before it is ready to merge.

**Comment by davrodpin** [comment]
Give me some time to improve the code and unit tests based on your feedbacks :smile:

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      DataSegment segment,
+      final int version,
+      final File compressedSegmentData,
+      final File descriptorFile,
+      final Map<String, String> azurePaths
+  )
+      throws StorageException, IOException, URISyntaxException
+  {
+    uploadThenDelete(compressedSegmentData, azurePaths.get("index"));
+    uploadThenDelete(descriptorFile, azurePaths.get("descriptor"));
```
Comment (nishantmonu51) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: why is upload and delete required ?

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureStorageContainer.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class AzureStorageContainer
+{
+
+  private final Logger log = new Logger(AzureStorageContainer.class);
```
Comment (nishantmonu51) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureStorageContainer.java]: make it static

**Comment by drcrallen** [comment]
@davrodpin : Can you please keep your Azure stuff in one PR instead of spreading them out? 

Whenever you're ready you can squash commits (see https://github.com/druid-io/druid/blob/master/CONTRIBUTING.md for short instructions on the recommended workflow)

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 2 above)
Comment (fjy) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: I think you may be looking for Files.createTempFile

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    CompressionUtils.zip(indexFilesDir, zipOutFile);
+
+    return zipOutFile;
+
+  }
+
+  public File createSegmentDescriptorFile(final ObjectMapper jsonMapper, final DataSegment segment) throws
+                                                                                                           IOException
+  {
+    File descriptorFile = File.createTempFile(TEMP_DIR, AzureStorageDruidModule.DESCRIPTOR_FILE_NAME);
```
Comment (fjy) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: same here, I think it should be Files.createTempFile

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 2 above)
Comment (fjy) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: are these tempfiles ever cleaned up?

Comment (fjy) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: ah okay, they appear to be cleaned up after uploading

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 3 above)
Comment (fjy) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: the files that are passed in are temp files that were created

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureStorageContainer.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  public List<String> emptyCloudBlobDirectory(final String path)
+      throws StorageException, URISyntaxException
+  {
+    List<String> deletedFiles = new ArrayList<>();
+
+    CloudBlobDirectory cloudBlobDirectory = cloudBlobContainer.getDirectoryReference(path);
+    for (ListBlobItem blobItem : cloudBlobDirectory.listBlobs()) {
+      if (blobItem instanceof CloudBlob) {
```
Comment (fjy) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureStorageContainer.java]: what happens if it is not? should we warn or throw an exception?

**Comment by fjy** [comment]
@davrodpin I had a few comments inline but overall looks pretty good. :+1: if we can clean up some of the code review comments

**Comment by davrodpin** [comment]
@drcrallen , @fjy and @nishantmonu51:  thanks for all the comments! I'm already working to address all the comments.

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 2 above)
Comment (davrodpin) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: > are these tempfiles ever cleaned up?

@fjy, I can simplify that using `deleteOnExit`, so the files will be deleted when the Peon's JVM terminates.

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 3 above)
Comment (davrodpin) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: > why is upload and delete required ?

I believe this is a common behavior to all Deep Storage implementations.
After creating those temporary files and uploading them to Azure, there is no use to keep a local copy of them.

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 5 above)
Comment (davrodpin) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: > same here, I think it should be Files.createTempFile

@fjy, I agree with @drcrallen that the way I am using the `File.createTempFile` causes confusion and I am going to fix that, but I don't see much difference of using `File.createTempFile` and `Files.createTempFile` on this specific case.

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureStorageContainer.java (hunk 6 above)
Comment (davrodpin) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureStorageContainer.java]: The code was written in that way because the files will be located under the given path directory and `cloudBlobDirectory.listBlobs()` can return files (`CloudBlob`) or virtual directories (`CloudBlobDirectory`).

I'll change the code to get a flat list of all files under the given path and then delete each one of them. If no file was deleted, a warn log message will be filed.

**Comment by davrodpin** [comment]
All feedback comments addressed, related documentation added and all commits squashed.

File: docs/content/Deep-Storage.md [hunk 7]
```diff
@@ -54,3 +54,17 @@ If you are using the Hadoop indexer, set your output directory to be a location
 
 Please note that this is a community contributed module and does not support Cassandra 2.x or hadoop-based batch indexing. For more information on using Cassandra as deep storage, see [Cassandra Deep Storage](Cassandra-Deep-Storage.html).
 
+## Azure
+
+[Microsoft Azure Storage](http://azure.microsoft.com/en-us/services/storage/) is another option for deep storage. This requires some additional druid configuration.
+
+```
+druid.storage.type=azure
```
Comment (drcrallen) [inline docs/content/Deep-Storage.md]: General note: The whole Deep Storage doc could use a reformat to use the property table stuff seen in other parts of the doc (not required for this PR)

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureLoadSpec.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+{
+  @JsonProperty
+  private final String storageDir;
+
+  private final AzureDataSegmentPuller puller;
+
+  @JsonCreator
+  public AzureLoadSpec(
+      @JacksonInject AzureDataSegmentPuller puller,
+      @JsonProperty("storageDir") String storageDir
```
Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureLoadSpec.java]: Most of the other load specs take in the full item name rather than just the directory. Is it possible to take in the full path here?

File: docs/content/Deep-Storage.md (hunk 7 above)
Comment (davrodpin) [inline docs/content/Deep-Storage.md]: @drcrallen: Something like this: https://github.com/druid-io/druid/pull/1289 ?

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureLoadSpec.java (hunk 8 above)
Comment (davrodpin) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureLoadSpec.java]: By item name you mean the file that it compress before upload (in this particular case, index.zip)?

Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureLoadSpec.java]: I mean the full path. If the azure spec requires a storage dir and a file name, that would be kind of like the amazon bucket and key. But hard-coding "index.zip" should be avoided if it can.

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+      final File descriptorFile,
+      final Map<String, String> azurePaths
+  )
+      throws StorageException, IOException, URISyntaxException
+  {
+    uploadThenDelete(compressedSegmentData, azurePaths.get("index"));
+    uploadThenDelete(descriptorFile, azurePaths.get("descriptor"));
+
+    return segment
+        .withSize(compressedSegmentData.length())
```
Comment (nishantmonu51) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: wouldn't length() always return 0 here, since the file has been deleted ?

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 3 above)
Comment (nishantmonu51) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: I wonder if you really need to create those temp files, or can use Azure BlobOutputStream as is done in HDFSDeepStorage.

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureLoadSpec.java (hunk 8 above)
Comment (davrodpin) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureLoadSpec.java]: @drcrallen, I've changed the code based on your feedback: https://github.com/druid-io/druid/pull/1305

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 9 above)
Comment (davrodpin) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: You're right. It has been fixed on https://github.com/druid-io/druid/pull/1305.
